teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

```diff
@@ -171,12 +201,503 @@ In addition to grouping related resources, Tekton Results provides long term sto
+            reason: Succeeded
+            status: "True"
+            type: Succeeded
+```
+
+#### Discussion
+
+While this approach is consistent with existing code, it does not comply with the
+[Kubernetes API conventions][subobjects] that recommend against maps.
+The [main problem][why-subobjects] with maps is:
+> The crux of maps is that it isn't clear to the user what "left-hand side strings"
+> are "magic keywords" in the config system/API vs. which are user data.
+
+### Beta API - Use Booleans
+
+We could add a behavior flag - `enable-full-embedded-status` - used to configure
+whether `PipelineRuns` should contain the embedded status or minimal status of
+its `TaskRuns` and `Runs`.
+
+Following our [policy][behavior-flags] on updating behavior flags:
+* The `enable-full-embedded-status` flag will be *true* by default.
+* After 9 months in v1beta1, the `enable-full-embedded-status` flag will be flipped
+  to *false* by default.
+* As soon as the next release in v1beta1, the `enable-full-embedded-status` flag will
+  be removed as well as the full embedded status fields.
+
+#### Discussion
+
+While the behavior flag taking booleans solves for the options we need, the 
+[Kubernetes API conventions][primitives] warn "think twice about boolean fields"
+because "many ideas start as boolean but eventually trend towards a small set of
+mutually exclusive options". In the small chance that users want to have both full
+and minimal embedded statuses as they make transitions, using a boolean for this
+field would be limiting. Therefore, we prefer the alternative to using string aliases.
+
+### Beta API - Default to Full then Both then Minimal
+
+To provide a smoother migration as required by our [API compatibility policy][deprecations]
+we will add a behavior flag - `embedded-status` - used to configure whether the
+`PipelineRuns` should contain:
+* the full embedded status of its `TaskRuns` and `Runs`
+* the minimal embedded status of its `TaskRuns`and `Runs`
+* both the full and minimal status of its `TaskRuns` and `Runs`
+
+Following our [policy][behavior-flags] on updating behavior flags:
+1. The `embedded-status` flag will be `full` by default, users can set it to `minimal`
+   or `both`.
+2. After 2 months in v1beta1, the `embedded-status` flag will be changed to `both`
+   by default, users can set it to `full` or `minimal`.
+3. After 7 more months in v1beta1, the `embedded-status` flag will be changed to `minimal`
+   by default, users can set it to `full` or `both`.
+4. As soon as the next release in v1beta1, `embedded-status` flag will be removed as
+   well as the full embedded status fields. In reality, this would take a bit longer
+   after confirming that users and contributors are ready for the flag to be removed.
+
+#### Discussion
+
+While this approach gives users more control by allowing them to receive both the full and
+minimal statuses, it causes more duplication and worsens the [problems](#motivation)
```

Comment:
+100

I really hope we won't do this!

Reply:
I think having a "both" option can be a very useful temporary option to ensure a seamless transition experience for Tekton API clients. The "both" approach is how we generally do backwards incompatible API changes to support older clients.

1. On release X, we have "both" as the default. CLI/Dashboard/any other clients have to rely on the old fields.

2. Operator updates pipeline to release X+1 where "minimal" is the default. CLI/Dashboard/any other clients are now all broken until they can be upgraded to a version that uses the new fields.

3. Because the only supported options are "full" and "minimal", operators have to ensure that every client is upgraded to a version that supports the new fields when they want to upgrade to version. 
One workaround for this would be for the clients to add in logic to support both the new and old fields at the same time but then by not adding the "both" option, we are essentially forcing our each of our clients to add the logic to support both versions.

4. If we did have a "both" option, the Tekton operator will flip the option to "both" for a short amount of time while allowing the clients to upgrade. Once all clients are upgraded, the operator will flip the switch to "minimal".